hledger-lib/Hledger/Utils/TH.hs

```diff
@@ -0,0 +1,54 @@
+module Hledger.Utils.TH
+( makeHledgerClassyLenses
+) where
+
+import Data.Char (toLower)
+import Data.List.Extra (uncons, unsnoc)
+import qualified Data.Set as Set
+import Language.Haskell.TH (DecsQ, Name, mkName, nameBase)
+import Lens.Micro ((&), (.~))
+import Lens.Micro.TH (DefName(TopName), lensClass, lensField, makeLensesWith, classyRules)
+
+-- | Make classy lenses for Hledger options fields.
+-- This is intended to be used with BalancingOpts, InputOpt, ReportOpts,
+-- ReportSpec, and CliOpts.
+-- When run on X, it will create a typeclass named HasX (except for ReportOpts,
+-- which will be named HasReportOptsNoUpdate) containing all the lenses for that type.
+-- If the field name starts with an underscore, the lens name will be created
+-- by stripping the underscore from the front on the name. If the field name ends with
+-- an underscore, the field name ends with an underscore, the lens name will be
+-- mostly created by stripping the underscore, but a few names for which this
+-- would create too many conflicts instead have a second underscore appended.
+-- ReportOpts fields for which updating them requires updating the query in
+-- ReportSpec are instead names by dropping the trailing underscore and
+-- appending NoUpdate to the name, e.g. querystring_ -> querystringNoUpdate.
```

Comment:
This is some tricky stuff we're adding, so I'm emphasising the docs. The above is excellent at explaining what this does. I think we could just add a sentence saying why we have these shenanigans - IIRC there are two reasons, 1. to temporarily accommodate legacy field names ending in underscore, postponing a big rename and 2. to assist with safely modifying ReportSpec, removing a source of bugs. And perhaps a link to the issue with the full discussion.